tfworker/util/copier.py

```diff
@@ -0,0 +1,217 @@
+
+    @property
+    def source(self):
+        """ source contains the source path providede """
+        return self._source
+
+    def get_destination(self, make_dir: bool = True, **kwargs) -> str:
+        """ get_destination returns the destination path, and optionally makes the destinatination directory """
+        if not (hasattr(self, "_destination") or "destination" in kwargs.keys()):
+            raise ValueError("no destination provided")
+        if "destination" in kwargs:
+            d = kwargs["destination"]
+        else:
+            d = self._destination
+
+        if make_dir:
+            make_d = Path(d)
+            make_d.mkdir(parents=True, exist_ok=True)
+
+        return d
+
+    def check_conflicts(self, path: str) -> None:
+        """ Checks for files with conflicting names in a path """
+        conflicting = []
+        if self.conflicts:
+            check_path = Path(path)
+            for check_file in check_path.glob("*"):
+                if check_file.name in self.conflicts:
+                    conflicting.append(check_file.name)
+
+        if conflicting:
+            raise FileExistsError(f"{','.join(conflicting)}")
+
+
+@CopyFactory.register("git")
+class GitCopier(Copier):
+    def copy(self, **kwargs) -> bool:
+        """ copy clones a remote git repo, and puts the requested files into the destination """
+        dest = self.get_destination(**kwargs)
+        branch = "master"
+
+        if "branch" in kwargs:
+            branch = kwargs["branch"]
+        if "sub_path" in kwargs:
+            sub_path = kwargs["sub_path"].strip("/")
+        else:
+            sub_path = ""
+
+        self.make_temp()
+        temp_path = f"{self._temp_dir}/{sub_path}"
+
+        pipe_exec(
+            f"git clone {self._source} --branch {branch} --single-branch ./",
+            cwd=self._temp_dir,
+        )
+
+        try:
+            self.check_conflicts(temp_path)
+        except FileExistsError as e:
+            self.clean_temp()
+            raise e
+
+        shutil.copytree(temp_path, dest, dirs_exist_ok=True)
+        self.clean_temp()
+
+        return True
+
+    @staticmethod
+    def type_match(source: str, **kwargs) -> bool:
+        """ type matches uses git to see if the source is a valid git remote """
+        try:
+            (return_code, _, _) = pipe_exec(f"git ls-remote {source}")
+        except (PermissionError, FileNotFoundError):
+            return False
+        if return_code == 0:
+            return True
+        return False
+
+    def make_temp(self) -> None:
+        if hasattr(self, "_temp_dir"):
+            pass
+        self._temp_dir = tempfile.mkdtemp()
+
+    def clean_temp(self) -> None:
+        """ clean_temp removes the temporary path used by this copier """
+        if hasattr(self, "_temp_dir"):
+            shutil.rmtree(self._temp_dir, ignore_errors=True)
+            del self._temp_dir
+
+
+@CopyFactory.register("fs")
+class FileSystemCopier(Copier):
```

Comment:
Does this support copying files outside of the `definition` path? e.g. I have a directory structure that looks something like:
```
file_I_want_to_copy
definitions/...
```

would this enable copying `file_I_want_to_copy` into my definition?

Reply:
Yup, it can copy from any valid path. To keep behavior consistent it accepts a"root_path" argument, which is equivalent to the workers "repository root" and then the source can be wherever. You can also omit the root_path entirely to use absolute path values.

Reply:
That will come in handy with the SDLF work in organizing the cloud formation templates.